### PR TITLE
Add Diplomat::Agent.self to get agent config

### PR DESCRIPTION
**Pull request by brugidou**

### CODE REVIEW

**Comment by brugidou** [comment]
I updated my commit

File: lib/diplomat/agent.rb [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+      url = ["/v1/agent/self"]
+
+      # If the request fails, it's probably due to a bad path
+      # so return a PathNotFound error.
+      begin
+        ret = @conn.get concat_url url
+      rescue Faraday::ClientError
+        raise Diplomat::PathNotFound
+      end
+      return JSON.parse(ret.body)
```
Comment (taharah) [inline lib/diplomat/agent.rb]: The return value does not match the documentation.

File: lib/diplomat/agent.rb [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+      url = ["/v1/agent/checks"]
+
+      # If the request fails, it's probably due to a bad path
+      # so return a PathNotFound error.
+      begin
+        ret = @conn.get concat_url url
+      rescue Faraday::ClientError
+        raise Diplomat::PathNotFound
+      end
+      return JSON.parse(ret.body)
```
Comment (taharah) [inline lib/diplomat/agent.rb]: The return value does not match the documentation.

File: lib/diplomat/agent.rb [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+      url = ["/v1/agent/services"]
+
+      # If the request fails, it's probably due to a bad path
+      # so return a PathNotFound error.
+      begin
+        ret = @conn.get concat_url url
+      rescue Faraday::ClientError
+        raise Diplomat::PathNotFound
+      end
+      return JSON.parse(ret.body)
```
Comment (taharah) [inline lib/diplomat/agent.rb]: The return value does not match the documentation.

File: lib/diplomat/agent.rb [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+      url = ["/v1/agent/members"]
+
+      # If the request fails, it's probably due to a bad path
+      # so return a PathNotFound error.
+      begin
+        ret = @conn.get concat_url url
+      rescue Faraday::ClientError
+        raise Diplomat::PathNotFound
+      end
+      return JSON.parse(ret.body)
```
Comment (taharah) [inline lib/diplomat/agent.rb]: The return value does not match the documentation.

File: spec/agent_spec.rb [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+require 'base64'
+
+describe Diplomat::Agent do
+
+  let(:faraday) { fake }
+
+  context "agent" do
+
+    it "self" do
+      json = <<EOF
```
Comment (taharah) [inline spec/agent_spec.rb]: Please change to using standard rspec variable assignment via `let` blocks.

File: spec/agent_spec.rb [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+      faraday.stub(:get).and_return(OpenStruct.new({ body: json }))
+
+      agent = Diplomat::Agent.new(faraday)
+
+      expect(agent.self["Config"]["NodeName"]).to eq("foobar")
+    end
+
+    it "checks" do
+      json = <<EOF
```
Comment (taharah) [inline spec/agent_spec.rb]: Please change to using standard rspec variable assignment via `let` blocks.

File: spec/agent_spec.rb [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+
+      faraday.stub(:get).and_return(OpenStruct.new({ body: json }))
+
+      agent = Diplomat::Agent.new(faraday)
+
+      expect(agent.checks["service:redis"]["Status"]).to eq("passing")
+    end
+
+	it "services" do
+      json = <<EOF
```
Comment (taharah) [inline spec/agent_spec.rb]: Please change to using standard rspec variable assignment via `let` blocks.

File: spec/agent_spec.rb [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+
+      faraday.stub(:get).and_return(OpenStruct.new({ body: json }))
+
+      agent = Diplomat::Agent.new(faraday)
+
+      expect(agent.services["redis"]["Port"]).to eq(8000)
+    end
+
+	it "members" do
+      json = <<EOF
```
Comment (taharah) [inline spec/agent_spec.rb]: Please change to using standard rspec variable assignment via `let` blocks.

**Comment by taharah** [comment]
RSpec test examples:
* https://github.com/WeAreFarmGeek/diplomat/commit/d9a9671c35b34eebe7fe5ea096d21f493c19c3fd
* https://github.com/WeAreFarmGeek/diplomat/commit/54314f7a496bf4fc456bf4c86f640eec2dbd601d

**Comment by brugidou** [comment]
Updated.

* I kept the raw JSON data and used `let(:json) {...}` because the JSON data comes from the official Consul documentation and thus is easier to maintain (see https://www.consul.io/docs/agent/http/agent.html )
* I realized that @johnhamelink asked me to add `checks`, `members` and `services` but checks and members are also available at `Diplomat::Check.checks`, `Diplomat::Members.get`. I am not sure how you want to organize the various classes/method according to the API hierarchy.

**Comment by taharah** [comment]
- Alright I can see why you kept the RAW JSON and that does make it easier to maintain as the Consul API evolves. It did take me a bit of time to convert the RAW JSON from the API documentation to a hash when I was writing tests. 
- Personally, I think the Diplomat gem should mirror the Consul API 1:1 as much as possible. This API endpoint has quite a bit of functionality and is probably the most in-depth endpoint. Thus, I do not mind adding the full functionality over time via several commits.

**Comment by maxvt** [comment]
In light of the most recent changes, is anything else needed to get this merged?

Being unable to read even the current DC has led us to implementing an Agent on our own, outside of Diplomat. Even having that single endpoint in would be a big help, I definitely agree with @taharah there is a benefit in even a partial support of the Agent API section.

**Comment by brugidou** [comment]
Hello, I don't undestand why this PR is marked as having "changes requested", I addressed all the reviewer's requests I believe.

**Comment by taharah** [comment]
@brugidou it has been merged. sorry for not responding for so long I didn't realize I did not have notifications setup.

**Pull request merged**